File: libs/moqtail-ts/src/model/control/publish.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { LengthExceedsMaxError } from '../error/error'
import { ControlMessageType } from './constant'
import { FullTrackName } from '../data'
import { MessageParameter, MessageParameters } from '../parameter/message_parameter'
import { TrackExtension, DeliveryTimeoutExtension } from '../extension_header/track_extension'
import { DeliveryTimeout } from '../parameter/message/delivery_timeout'
import { Forward } from '../parameter/message/forward'

export class Publish {
  constructor(
    public readonly requestId: bigint,
    public readonly fullTrackName: FullTrackName,
    public readonly trackAlias: bigint,
    public readonly parameters: MessageParameter[],
    public readonly trackExtensions: TrackExtension[] = [],
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.Publish
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.Publish)
    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putBytes(this.fullTrackName.serialize().toUint8Array())
    payload.putVI(this.trackAlias)
    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putKeyValuePair(param.toKeyValuePair())
    }
    TrackExtension.serializeInto(this.trackExtensions, payload)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('Publish::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): Publish {
    const requestId = buf.getVI()
    const fullTrackName = buf.getFullTrackName()
    const trackAlias = buf.getVI()
    const paramCount = buf.getNumberVI()
    const rawParams = new Array(paramCount)
    for (let i = 0; i < paramCount; i++) {
      rawParams[i] = buf.getKeyValuePair()
    }
    const parameters = MessageParameters.fromKeyValuePairs(rawParams)
    const trackExtensions = TrackExtension.deserializeAll(buf)
    return new Publish(requestId, fullTrackName, trackAlias, parameters, trackExtensions)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('Publish', () => {
    test('roundtrip', () => {
      const requestId = 12345n
      const fullTrackName = FullTrackName.tryNew('video/stream', 'camera1')
      const trackAlias = 123n
      const parameters = [new DeliveryTimeout(500n), new Forward(true)]
      const publish = new Publish(requestId, fullTrackName, trackAlias, parameters)

      const frozen = publish.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Publish))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = Publish.parsePayload(frozen)
      expect(deserialized.requestId).toBe(publish.requestId)
      expect(deserialized.fullTrackName.namespace.equals(publish.fullTrackName.namespace)).toBe(true)
      expect(deserialized.fullTrackName.name).toEqual(publish.fullTrackName.name)
      expect(deserialized.trackAlias).toBe(publish.trackAlias)
      expect(deserialized.parameters.length).toBe(2)
      expect(deserialized.trackExtensions.length).toBe(0)
      expect(frozen.remaining).toBe(0)
    })

    test('roundtrip no parameters', () => {
      const requestId = 12345n
      const fullTrackName = FullTrackName.tryNew('video/stream', 'camera1')
      const trackAlias = 123n
      const publish = new Publish(requestId, fullTrackName, trackAlias, [])

      const frozen = publish.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Publish))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = Publish.parsePayload(frozen)
      expect(deserialized.requestId).toBe(publish.requestId)
      expect(deserialized.parameters.length).toBe(0)
      expect(frozen.remaining).toBe(0)
    })

    test('roundtrip with track extensions', () => {
      const requestId = 12345n
      const fullTrackName = FullTrackName.tryNew('video/stream', 'camera1')
      const publish = new Publish(requestId, fullTrackName, 1n, [], [new DeliveryTimeoutExtension(3000n)])
      const frozen = publish.serialize()
      frozen.getVI() // message type
      const msgLength = frozen.getU16()
      const payload = new FrozenByteBuffer(frozen.getBytes(msgLength))
      const deserialized = Publish.parsePayload(payload)
      expect(deserialized.trackExtensions.length).toBe(1)
      expect(deserialized.trackExtensions[0]).toBeInstanceOf(DeliveryTimeoutExtension)
      expect(payload.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const requestId = 12345n
      const fullTrackName = FullTrackName.tryNew('video/stream', 'camera1')
      const trackAlias = 123n
      const parameters = [new DeliveryTimeout(500n)]
      const publish = new Publish(requestId, fullTrackName, trackAlias, parameters)
      const serialized = publish.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Publish))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const payload = new FrozenByteBuffer(frozen.getBytes(msgLength))
      const deserialized = Publish.parsePayload(payload)
      expect(deserialized.requestId).toBe(publish.requestId)
      expect(deserialized.fullTrackName.namespace.equals(publish.fullTrackName.namespace)).toBe(true)
      expect(deserialized.fullTrackName.name).toEqual(publish.fullTrackName.name)
      expect(payload.remaining).toBe(0)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const requestId = 12345n
      const fullTrackName = FullTrackName.tryNew('video/stream', 'camera1')
      const publish = new Publish(requestId, fullTrackName, 123n, [new DeliveryTimeout(500n)])
      const serialized = publish.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        Publish.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/publish.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::tuple::{Tuple, TupleField};
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use crate::model::parameter::message_parameter::{
  MessageParameter, deserialize_message_parameters, serialize_message_parameters,
};
use crate::model::property::track_property::{
  TrackProperty, deserialize_track_properties, serialize_track_properties,
};
use bytes::{Buf, BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct Publish {
  pub request_id: u64,
  pub track_namespace: Tuple,
  pub track_name: TupleField,
  pub track_alias: u64,
  pub parameters: Vec<MessageParameter>,
  pub track_properties: Vec<TrackProperty>,
}

impl Publish {
  pub fn new(
    request_id: u64,
    track_namespace: Tuple,
    track_name: TupleField,
    track_alias: u64,
    parameters: Vec<MessageParameter>,
    track_properties: Vec<TrackProperty>,
  ) -> Self {
    Self {
      request_id,
      track_namespace,
      track_name,
      track_alias,
      parameters,
      track_properties,
    }
  }
}

impl ControlMessageTrait for Publish {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::Publish)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;
    payload.extend_from_slice(&self.track_namespace.serialize()?);

    // Track Name Length and Track Name
    payload.put_vi(self.track_name.len() as u64)?;
    payload.extend_from_slice(self.track_name.as_bytes());

    payload.put_vi(self.track_alias)?;

    // Parameters
    payload.put_vi(self.parameters.len() as u64)?;
    payload.extend_from_slice(&serialize_message_parameters(&self.parameters)?);

    // Track Properties (no length prefix; bounded by outer message Length field)
    payload.extend_from_slice(&serialize_track_properties(&self.track_properties)?);

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Publish::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;
    let track_namespace = Tuple::deserialize(payload)?;

    let track_name_length = payload.get_vi()? as usize;
    if payload.remaining() < track_name_length {
      return Err(ParseError::NotEnoughBytes {
        context: "Publish::parse_payload(track_name)",
        needed: track_name_length,
        available: payload.remaining(),
      });
    }
    let track_name = TupleField::new(payload.copy_to_bytes(track_name_length));

    let track_alias = payload.get_vi()?;

    let param_count = payload.get_vi()?;
    let parameters =
      deserialize_message_parameters(payload, param_count, ControlMessageType::Publish)?;

    // Track Properties: consume whatever remains in the payload
    let track_properties = deserialize_track_properties(payload)?;

    Ok(Box::new(Publish {
      request_id,
      track_namespace,
      track_name,
      track_alias,
      parameters,
      track_properties,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::Publish
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::model::common::location::Location;
  use crate::model::control::constant::GroupOrder;
  use crate::model::parameter::message_parameter::MessageParameter;
  use crate::model::property::track_property::TrackProperty;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let mut publish = Publish::new(
      123,
      Tuple::from_utf8_path("example/track"),
      TupleField::from_utf8("video"),
      456,
      vec![
        MessageParameter::new_group_order(GroupOrder::Ascending),
        MessageParameter::new_largest_object(Location::new(10, 20)),
        MessageParameter::Forward { forward: true },
        MessageParameter::new_expires(1000),
      ],
      vec![],
    );
    // Wire encoding canonicalizes parameter order ascending by type (delta-encoding requirement).
    publish.parameters.sort_by_key(|p| p.type_value());

    let mut buf = publish.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Publish as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = Publish::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_roundtrip_no_parameters() {
    let publish = Publish::new(
      123,
      Tuple::from_utf8_path("example/track"),
      TupleField::from_utf8("video"),
      456,
      vec![],
      vec![],
    );

    let mut buf = publish.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Publish as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = Publish::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_roundtrip_with_track_properties() {
    let publish = Publish::new(
      1,
      Tuple::from_utf8_path("ns/track"),
      TupleField::from_utf8("audio"),
      7,
      vec![],
      vec![
        TrackProperty::DeliveryTimeout { timeout_ms: 3000 },
        TrackProperty::DefaultPublisherPriority { priority: 100 },
      ],
    );

    let mut buf = publish.serialize().unwrap();
    let _msg_type = buf.get_vi().unwrap();
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = Publish::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish);
    assert!(!buf.has_remaining());
  }
}
